### PR TITLE
Alert for blocked ESP protocol as part of diagnose

### PR DESCRIPTION
**Pull request by sridhargaddam**
When there is no NAT between the Gateway nodes and Libreswan cable-driver is used, IPsec protocol uses UDP/4500 port for exchanging the keys and ESP for data transfer. As part of subctl diagnose command, this PR now validates if there is a potential issue with ESP and returns appropriate error message.

### CODE REVIEW

File: pkg/diagnose/firewall.go [hunk 1]
```diff
@@ -245,22 +251,49 @@ func verifyConnectivity(localClusterInfo, remoteClusterInfo *cluster.Info, names
...
 	}
 
-	if options.VerboseOutput {
-		status.Success("tcpdump output from sniffer pod on Gateway node:\n%s", sPod.PodOutput)
-	}
+	return nil
+}
 
+func validateOutput(sPod *pods.Scheduled, clientMessage, hostname string, destPort int32,
+	noNatWithIPsec bool, status reporter.Interface,
```
Comment (skitt) [inline pkg/diagnose/firewall.go]: It’s not obvious what the connection between `noNatWithIPsec` and ESP is — I think this function would be clearer if its parameter name reflected the use within the function:

```suggestion
	espNeeded bool, status reporter.Interface,
```

**Review by skitt — CHANGES_REQUESTED** [review]

File: pkg/diagnose/firewall.go [hunk 2]
```diff
@@ -245,22 +251,49 @@ func verifyConnectivity(localClusterInfo, remoteClusterInfo *cluster.Info, names
...
+			return status.Error(fmt.Errorf("the tcpdump output from the sniffer pod does not include the message"+
+				" sent from client pod. Please check that your firewall configuration allows UDP/%d traffic"+
+				" and ESP traffic on the %q nodes. Actual pod output: \n%s", destPort, hostname,
+				truncate(sPod.PodOutput)), "")
+		}
+
 		return status.Error(fmt.Errorf("the tcpdump output from the sniffer pod does not include the message"+
 			" sent from client pod. Please check that your firewall configuration allows UDP/%d traffic"+
-			" on the %q node. Actual pod output: \n%s", destPort, localEndpoint.Spec.Hostname, truncate(sPod.PodOutput)), "")
+			" on the %q node. Actual pod output: \n%s", destPort, hostname, truncate(sPod.PodOutput)), "")
```
Comment (tpantelis) [inline pkg/diagnose/firewall.go]:
```
espMsg := "'
If espNeeded {
     espMsg = "and ESP traffic"
}

return status.Error(fmt.Errorf("the tcpdump output from the sniffer pod does not include the message"+
			" sent from client pod. Please check that your firewall configuration allows UDP/%d traffic"+
			"%s on the %q node. Actual pod output: \n%s", destPort, espMsg, hostname, truncate(sPod.PodOutput)), "")

```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/diagnose/firewall.go [hunk 3]
```diff
@@ -245,22 +251,48 @@ func verifyConnectivity(localClusterInfo, remoteClusterInfo *cluster.Info, names
...
+}
 
+func validateOutput(sPod *pods.Scheduled, clientMessage, hostname string, destPort int32,
+	espNeeded bool, status reporter.Interface,
+) error {
 	if !strings.Contains(sPod.PodOutput, clientMessage) {
+		espMsg := ""
+
+		if espNeeded {
+			espMsg = "and ESP traffic"
```
Comment (skitt) [inline pkg/diagnose/firewall.go]:
```suggestion
			espMsg = " and ESP traffic"
```

**Review by skitt — CHANGES_REQUESTED** [review]
Thanks, a minor tweak still to ensure that “traffic” and “and” don’t end up stuck together.

**Review by tpantelis — APPROVED** [review]

**Review by skitt — APPROVED** [review]

**Comment by sridhargaddam** [comment]
Release note PR: https://github.com/submariner-io/submariner-website/pull/1010

**Pull request merged**